proto/p4/v1/p4runtime.proto

```diff
@@ -210,6 +210,28 @@ message TableEntry {
   IdleTimeout time_since_last_hit = 10;
   // Arbitrary metadata from the controller that is opaque to the target.
   bytes metadata = 11;
+  // const is true only for entries that cannot be deleted or
+  // modified, e.g. those in a table with `has_initial_entries = true`
+  // that have been marked `const` _on that individual entry_ in the
+  // P4 source code within its `entries` table property.
+  //
+  // TODO: Should const be true for all entries read from a table
+  // declared with `const entries`?  Would that be considered a
```

Comment:
Existing implementations will not populate this field, evaluating to 0=false, even for constant entries, resulting in bogus information. I don't know if this constitutes "backwards incompatibility," or it simply means existing implementations are suddently non-conformant with a new requirement. If the consensus is to adopt the proposal despite this fact, we would need at a minimum to write an advisory that this field should be ignored unless the implementation conforms with the spec version which adopts this.

Reply:
That does seem like an undesirable state.  One way to avoid such a situation is to define that when reading entries from a table with `const entries`, or from a default entry declared with `const`, that this new TableEntry `const` flag will always be false, to ensure that old and new servers return the same result for these kinds of read requests both before and after this spec change is made.

If we did that, we would try to document this as clearly as we can, and the reason why (because it would not be immediately obvious to someone new to the P4Runtime spec a year or two from now), and that the only time the server would set the `const` flag to `true` in a response to a read request would be for a table that had a `const` flag on an entry in a table with this new style of initial entries.

Reply:
Your logic is sound, but it doesn't feel satisfying or intuitive overall - too many conditions to wrap one's head around. It seems like a long-term solution to a short-term problem. Perhaps we should get others to weigh in on this. I'm not sure how objectionable it is - clients of this older server could simply not try to use the field until the server supports it, same as any other new feature. Perhaps we should get others to weigh in on this.

Reply:
I would love to get others to weigh in on this.  @smolkaj @jonathan-dilorenzo @thantry Can you or someone you know at Google who is interested in getting initial entries actually working in an implementation (as opposed to merely specified in P4.org documents) please take a look, or get someone else to take a look, at this PR in detail and comment?

Reply:
I would still like others to weigh in on this question, but thinking about this some more, it does seem in the spirit of Protobuf based messages to consider it still 'backwards compatible' if an old message has a new field with additional information added to it.  If we have a new `const` field in `TableEntry` messages, and new P4Runtime servers in the future return `true` for this field for entries declared with `const entries` in the source code, even though older P4Runtime servers return no such field at all, that seems like something that P4Runtime API clients should just need to handle gracefully.

Reply:
With commit 12 on this PR, I have changed the proposed comment being commented on here to say that this new `const` field for TableEntry messages WILL be set to true by the server for all entries read from a table with `const entries`, and also for reading the default action from a table that has `const default_action`.  We can change that before approving, of course, depending on whether anyone else has arguments against this choice.

Reply:
Sorry for missing this until now.

I am not someone who is planning to immediately use initial entries at this point, but can answer from a protobuf / semantic versioning perspective.

This requires a MINOR version bump in terms of https://semver.org/.

This is a "backward compatible API change" in the sense that an "old client" (using the old .proto) will continue to work without problems with a "new server" who adopt the new API:
* READ: When an old client gets a message with the new field (in binary format), protobuf will simply ignore the extra data and the client will not see it. 
* WRITE: When an old client writes a message without the new field (in binary format), a new server will see the field as being set to `false`. That works out fine for this proposal.

Note that a new client won't automagically work with an old server, as mentioned by other above: they will read back all entries with `is_const: false` even if the entries are const. But since the client made a deliberate choice here to adopt the new API, it's also their job to ensure they can support older servers as well, if they wish. ("This is a backward compatible API change" means that upgrading servers to the new API won't break clients -- it doesn't talk about what happens if clients start using the new API with servers that don't support it...)

@jafingerhut Perhaps we should leave a comment saying that the field will be set by P4Runtime servers starting at version x.y.z?

Reply:
Version x.y.z of what project?  Stratum?  p4lang/PI?  The P4Runtime API specification?

Reply:
I added two "Added in 1.4.0" comments with my latest commit to this PR, one for `is_const` in `TableEntry`, and the other for `has_initial_entries` in `Table`.